Sources/NIOHTTPCompression/HTTPCompression.swift

```diff
@@ -0,0 +1,130 @@
+            // deflateBound() provides an upper limit on the number of bytes the input can
+            // compress to. We add 5 bytes to handle the fact that Z_SYNC_FLUSH will append
+            // an empty stored block that is 5 bytes long.
+            let bufferSize = Int(deflateBound(&stream, UInt(inputBuffer.readableBytes)))
+            var outputBuffer = allocator.buffer(capacity: bufferSize)
+            stream.oneShotDeflate(from: &inputBuffer, to: &outputBuffer, flag: flags)
+            return outputBuffer
+        }
+        
+        mutating func shutdown() {
+            assert(isActive)
+            isActive = false
+            deflateEnd(&stream)
+        }
+        
+        mutating func shutdownIfActive() {
+            if isActive {
+                isActive = false
+                deflateEnd(&stream)
+            }
+        }
+    }
+}
+
+extension z_stream {
+    /// Executes deflate from one buffer to another buffer. The advantage of this method is that it
+    /// will ensure that the stream is "safe" after each call (that is, that the stream does not have
+    /// pointers to byte buffers any longer).
+    mutating func oneShotDeflate(from: inout ByteBuffer, to: inout ByteBuffer, flag: Int32) {
+        defer {
+            self.avail_in = 0
+            self.next_in = nil
+            self.avail_out = 0
+            self.next_out = nil
+        }
+
+        from.readWithUnsafeMutableReadableBytes { dataPtr in
+            let typedPtr = dataPtr.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            let typedDataPtr = UnsafeMutableBufferPointer(start: typedPtr,
+                                                          count: dataPtr.count)
+
+            self.avail_in = UInt32(typedDataPtr.count)
+            self.next_in = typedDataPtr.baseAddress!
+
+            let rc = deflateToBuffer(buffer: &to, flag: flag)
+            precondition(rc == Z_OK || rc == Z_STREAM_END, "One-shot compression failed: \(rc)")
+
+            return typedDataPtr.count - Int(self.avail_in)
+        }
+    }
+
+    /// A private function that sets the deflate target buffer and then calls deflate.
+    /// This relies on having the input set by the previous caller: it will use whatever input was
+    /// configured.
+    private mutating func deflateToBuffer(buffer: inout ByteBuffer, flag: Int32) -> Int32 {
+        var rc = Z_OK
+
+        buffer.writeWithUnsafeMutableBytes(minimumWritableBytes: buffer.capacity) { outputPtr in
+            let typedOutputPtr = UnsafeMutableBufferPointer(start: outputPtr.baseAddress!.assumingMemoryBound(to: UInt8.self),
+                                                            count: outputPtr.count)
+            self.avail_out = UInt32(typedOutputPtr.count)
+            self.next_out = typedOutputPtr.baseAddress!
+            rc = deflate(&self, flag)
+            return typedOutputPtr.count - Int(self.avail_out)
+        }
+
+        return rc
+    }
+}
```

Comment:
The fact that this extension moved between files makes it really hard to review if there are any meaningful differences between the two implementations. Can we restore the extension to its original location and remove the `private`?